File: hash/javascript/sha1_counter.js
#!/usr/bin/node

var BigNumber = require('bignumber.js');
var crypto = require('crypto');
var fs = require('fs');
var stdio = require('stdio');

var opts = stdio.getopt({
    'binary': {key: 'b', args: 0, description: 'Raw binary output.'},
    'key': {key: 'k', args: 1, description: 'SHA-1 key.'},
    'numbers': {key: 'n', args: 1, description: 'Quantity of random numbers.'}
});

function xor(s1, s2) {
    s1 = new Buffer(s1);
    s2 = new Buffer(s2);
    var xor = [];
    for (var i=0; i<s1.length; i++) {
        xor.push(s1[i]^s2[i]);
    }
    return new Buffer(xor);
}

function digest(s) {
    var sha1 = crypto.createHash('sha1');
    return sha1.update(s).digest();
}

var data = fs.readFileSync('/proc/interrupts').toString();
var d = crypto.createHash('sha1');
var sha = d.update(data).digest();
var d = crypto.createHash('ripemd160');
var ripemd = d.update(data).digest();
var salt = xor(sha, ripemd);
var key = crypto.pbkdf2Sync(sha, salt, 0, 16);
var seed = crypto.pbkdf2Sync(ripemd, salt, 0, 16);
var number = 1;

if(opts.key) key = crypto.pbkdf2Sync(opts.key, salt, 0, 16);
if(opts.numbers) number = opts.numbers;

// Almost never return exponential notation
BigNumber.config({ EXPONENTIAL_AT: 1e+9 });

var counter = Date.now();

for (var i=0; i<number; i++) {
    var out = new Buffer(digest(counter+key));
    counter++;
    if(opts.binary) process.stdout.write(out);
    else console.log(new BigNumber(out.toString('hex'), 16).toString());
}

File: hash/python/sha1_counter.py
#!/usr/bin/python

import argparse
import struct
import sys
import time
from Crypto.Hash import RIPEMD, SHA
from Crypto.Protocol import KDF

parser = argparse.ArgumentParser(description='SHA-1 in counter mode CSPRNG')
parser.add_argument('-b','--binary', help='Raw binary output.', action="store_true")
parser.add_argument('-k','--key', help='SHA-1 key.')
parser.add_argument('-n','--numbers', help='Quantity of random numbers.')
args = parser.parse_args()

def sxor(s1, s2):
    return ''.join(chr(ord(a)^ord(b)) for a,b in zip(s1,s2))

with open('/proc/interrupts','r') as f:
    data = f.read().replace('\n','')

d = RIPEMD.new(data)
ripemd = d.hexdigest()
d = SHA.new(data)
sha = d.hexdigest()
salt = sxor(sha, ripemd)
key = KDF.PBKDF2(sha, salt, 16, 0)
number = 1

if args.key:
    key = KDF.PBKDF2(bytes(args.key), salt, 16, 0)

if args.numbers:
    number = int(args.numbers)

counter = int(time.time())

# SHA-1 in counter mode
for i in xrange(0, number):
    sha1 = SHA.new(key)
    sha1.update(bytes(counter))
    out = sha1.digest()
    counter += 1
    if args.binary:
        sys.stdout.write(out)
    else:
        res = struct.unpack_from('QQ', out)
        print(res[0]*2**64+res[1])
